### PR TITLE
Randomized tests on storage cache

### PR DESCRIPTION
**Pull request by NikVolf**
along with some extra test cases

### CODE REVIEW

**Comment by NikVolf** [comment]
Not sure such tests should be applied on the cache level or on the state storage + cache level though, @arkpar

**Comment by arkpar** [comment]
Unit tests are fine with in-memory backend I suppose. We could add some tests with pre-populated backend and some that simulate backend changes.

**Comment by gavofyork** [comment]
test failures...

File: core/client/db/src/storage_cache.rs [hunk 1]
```diff
@@ -765,3 +833,490 @@ mod tests {
 		assert_eq!(s.storage(&key).unwrap(), None);
 	}
 }
+
+#[cfg(test)]
+mod qc {
+	use std::collections::{HashMap, hash_map::Entry};
+
+    use quickcheck::{quickcheck, TestResult, Arbitrary};
```
Comment (kianenigma) [inline core/client/db/src/storage_cache.rs]:
```suggestion
	use quickcheck::{quickcheck, TestResult, Arbitrary};
```

**Comment by NikVolf** [comment]
> Unit tests are fine with in-memory backend I suppose. We could add some tests with pre-populated backend and some that simulate backend changes.

Yep, the same api can be generalised and reused there as well!

**Comment by NikVolf** [comment]
> test failures...

fixed

File: core/client/db/src/storage_cache.rs [hunk 2]
```diff
@@ -654,6 +666,62 @@ mod tests {
 		assert!(s.storage(&key).unwrap().is_none());
 	}
 
+	#[test]
+	fn simple_fork() {
+		//init_log();
```
Comment (kianenigma) [inline core/client/db/src/storage_cache.rs]:
```suggestion

```

File: core/client/db/src/storage_cache.rs [hunk 3]
```diff
@@ -654,6 +666,62 @@ mod tests {
...
+		let s = CachingState::new(InMemory::<Blake2Hasher>::default(), shared.clone(), Some(h2a.clone()));
+		assert_eq!(s.storage(&key).unwrap().unwrap(), vec![2]);
+	}
+
+	#[test]
+	fn double_fork() {
+		//init_log();
+		let root_parent = H256::random();
+		let key = H256::random()[..].to_vec();
+		let h1 = H256::random();
```
Comment (kianenigma) [inline core/client/db/src/storage_cache.rs]: This is a nit that you can ignore but I would def. consider if I can hide this process of generating `N` random `h256::random()`s somewhere. In the easiest case, if you are fine with using them behind an index (as opposed to h1, h1a etc.) you can do something along the lines of `(0..N).map(H256::random).collect()` or sth similar.

Comment (NikVolf) [reply core/client/db/src/storage_cache.rs]: it will be hard to semantically distinguish forks and so on but thanks!

**Pull request merged**